core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +70,206 @@ class SYCLInternal {
+          throw std::bad_alloc();
+        }
+
+        m_capacity = n;
+      }
+
+      return m_capacity;
+    }
+
+   private:
+    // This will memcpy an object T into memory held by this object
+    // returns: a T* to that object
+    //
+    // Note:  it is UB to dereference this pointer with an object that is
+    // not an implicit-lifetime nor trivially-copyable type, but presumably much
+    // faster because we can use USM device memory
+    template <typename T>
+    std::unique_ptr<T, Deleter> memcpy_from(const T& t) {
+      reserve(sizeof(T));
+      sycl::event memcopied = m_q->memcpy(m_data, std::addressof(t), sizeof(T));
+      memcopied.wait();
+
+      std::unique_ptr<T, Deleter> ptr(reinterpret_cast<T*>(m_data),
+                                      Deleter(this));
+      m_size = sizeof(T);
+      return ptr;
+    }
+
+    // This will copy-constuct an object T into memory held by this object
+    // returns: a unique_ptr<T, destruct_delete> that will call the
+    // destructor on the type when it goes out of scope.
+    //
+    // Note:  This will not work with USM device memory
+    template <typename T>
+    std::unique_ptr<T, Deleter> copy_construct_from(const T& t) {
+      static_assert(kind != sycl::usm::alloc::device,
+                    "Cannot copy construct into USM device memory");
+
+      reserve(sizeof(T));
+
+      std::unique_ptr<T, Deleter> ptr(new (m_data) T(t), Deleter(this));
+      m_size = sizeof(T);
+      return ptr;
+    }
+
+   public:
+    // Performs either memcpy (for USM device memory) and returns a T*
+    // (but is technically UB when dereferenced on an object that is not
+    // an implicit-lifetime nor trivially-copyable type
+    //
+    // or
+    //
+    // performs copy construction (for other USM memory types) and returns a
+    // unique_ptr<T, ...>
+    template <typename T>
+    std::unique_ptr<T, Deleter> copy_from(const T& t) {
+      if constexpr (sycl::usm::alloc::device == kind)
+        return memcpy_from(t);
+      else
+        return copy_construct_from(t);
+    }
+
+   private:
+    template <typename T>
+    T& memcpy_to(T& t) {
+      assert(sizeof(T) == m_size);
+
+      sycl::event memcopied = m_q->memcpy(std::addressof(t), m_data, sizeof(T));
+      memcopied.wait();
+
+      return t;
+    }
+
+    template <typename T>
+    T& move_assign_to(T& t) {
+      static_assert(kind != sycl::usm::alloc::device,
+                    "Cannot move_assign_to from USM device memory");
+
+      assert(sizeof(T) == m_size);
+
+      t = std::move(*static_cast<T*>(m_data));
+
+      return t;
+    }
+
+   public:
+    template <typename T>
+    T& transfer_to(T& t) {
+      if constexpr (sycl::usm::alloc::device == kind)
+        return memcpy_to(t);
+      else
+        return move_assign_to(t);
+    }
+
+   private:
+    // USMObjectMem class invariants
+    //
+    //  !m_data == !m_capacity
+    //  m_q || !m_data
+    //  m_data || !m_size
+    //  m_size <= m_capacity
```

Comment:
Wasn't  sure how to read your invariants before I read the comment below.  Maybe precede with "all four expressions below must evaluate to `true`"?

Reply:
Improved.